ignite/contrib/handlers.py

```diff
@@ -0,0 +1,87 @@
+try:
+    from tqdm import tqdm
+except ImportError:
+    raise RuntimeError("This contrib module requires tqdm to be installed")
+
+from ignite.engine import Events
+
+
+class ProgressBar:
+    """
+    TQDM progress bar handler to log training progress and computed metrics
+
+    Args:
+        engine: ignite.Engine object
+        loader: iterable or dataloader object
+        output_transform: transform a function that transforms engine.state.output
+                into a dictionary of format {name: value}
+        mode (str): 'iteration' or 'epoch' (default=epoch)
+        log_interval (int): interval of which the metrics information is displayed (default=1)
+
+    Example:
+        (...)
+        pbar = ProgressBar(trainer, train_loader, output_transform=lambda x: {'loss': x})
+        trainer.add_handler(Events.ITERATION_COMPLETED, pbar)
+    """
+
+    def __init__(self, engine, loader, output_transform=lambda x: x, mode='epoch', log_interval=1):
+        self.num_iterations = len(loader)
```

Comment:
Coming back to this one. So I run the following code: 
```
trainer = Engine(process_function)
pbar = ProgressBar(trainer, train_loader)
```
by default its epoch mode but I see:
```
[0/105]   0%|           [00:00<?]
```
so 105 is number of batches

Reply:
that's intentional, because the mode only affects the `tqdm.write()` operation

Reply:
Otherwise, here is the code I'm running in the notebook
```python
from __future__ import print_function
from argparse import ArgumentParser

from torch import nn
from torch.optim import SGD
from torch.utils.data import DataLoader
import torch
import torch.nn.functional as F
from torchvision.transforms import Compose, ToTensor, Normalize
from torchvision.datasets import FakeData
from torchvision.models.resnet import resnet34

from ignite.engine import Events, Engine, _prepare_batch
from ignite.metrics import CategoricalAccuracy, Loss, TopKCategoricalAccuracy

from ignite.contrib.handlers import ProgressBar


def get_data_loaders(train_batch_size, val_batch_size):
    data_transform = Compose([ToTensor()])

    train_loader = DataLoader(FakeData(size=10000, image_size=(3, 224, 224), num_classes=10, 
                                       transform=data_transform, target_transform=lambda t: t.long()),
                              batch_size=train_batch_size, shuffle=True)

    val_loader = DataLoader(FakeData(size=10000, image_size=(3, 224, 224), num_classes=10, 
                                     transform=data_transform, target_transform=lambda t: t.long()),
                            batch_size=val_batch_size, shuffle=False)
    return train_loader, val_loader


def run(train_batch_size, val_batch_size, epochs, lr, momentum, log_interval):
    train_loader, _ = get_data_loaders(train_batch_size, val_batch_size)
    model = resnet34(pretrained=False, num_classes=10)
    device = 'cpu'

    if torch.cuda.is_available():
        device = 'cuda'

    loss_fn = F.cross_entropy
    optimizer = SGD(model.parameters(), lr=lr, momentum=momentum)
    
    if device:
        model.to(device)

    def process_function(engine, batch):
        model.train()
        optimizer.zero_grad()
        x, y = _prepare_batch(batch, device=device)
        y_pred = model(x)
        loss = loss_fn(y_pred, y)
        loss.backward()
        optimizer.step()
        return [loss.item(), y_pred, y]

    trainer = Engine(process_function)
    
    pbar = ProgressBar(trainer, train_loader)
    
    output_transform = lambda x: [x[1], x[2]]
    metrics={
        'accuracy': CategoricalAccuracy(output_transform=output_transform),
        'topk': TopKCategoricalAccuracy(output_transform=output_transform),
        'xentropy': Loss(loss_fn, output_transform=output_transform)
    }
    for name, metric in metrics.items():
        metric.attach(trainer, name)


    @trainer.on(Events.ITERATION_COMPLETED)
    def log_training_loss(engine):
        iter = (engine.state.iteration - 1) % len(train_loader) + 1
#         if iter % log_interval == 0:
#             print("Epoch[{}] Iteration[{}/{}] Loss: {:.2f}"
#                   "".format(engine.state.epoch, iter, len(train_loader), engine.state.output[0]))

    trainer.run(train_loader, max_epochs=epochs)
    print(trainer.state.metrics)
```
and I get 
```
[0/105]   0%|           [00:00<?]
Epoch 1
[0/105]   0%|           [00:00<?]
Epoch 2
[0/105]   0%|           [00:00<?]
Epoch 3
[0/105]   0%|           [00:00<?]
Epoch 4
```

Reply:
you need to pass the `output_transform` argument:

Not sure what is the format in your code, but this transform is supposed to return a dictionary in the format `{'metric_name1': metric_value1, 'metric_name2': metric_value2, (...)}`

Reply:
you can even do: `lambda x: {'xentropy': engine.state.metrics['xentropy']}`

Reply:
also you need to add: 
`trainer.add_handler(Events.ITERATION_COMPLETED, pbar)`

Reply:
Okay, let's say the previous example is rather specific. 
Here is another test and I do not understand the intension to the output.
```python
from __future__ import print_function
from argparse import ArgumentParser

from torch import nn
from torch.optim import SGD
from torch.utils.data import DataLoader
import torch
import torch.nn.functional as F
from torchvision.transforms import Compose, ToTensor, Normalize
from torchvision.datasets import MNIST

from ignite.engine import Events, create_supervised_trainer, create_supervised_evaluator
from ignite.metrics import CategoricalAccuracy, Loss

from ignite.contrib.handlers import ProgressBar


class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = nn.Conv2d(1, 10, kernel_size=5)
        self.conv2 = nn.Conv2d(10, 20, kernel_size=5)
        self.conv2_drop = nn.Dropout2d()
        self.fc1 = nn.Linear(320, 50)
        self.fc2 = nn.Linear(50, 10)

    def forward(self, x):
        x = F.relu(F.max_pool2d(self.conv1(x), 2))
        x = F.relu(F.max_pool2d(self.conv2_drop(self.conv2(x)), 2))
        x = x.view(-1, 320)
        x = F.relu(self.fc1(x))
        x = F.dropout(x, training=self.training)
        x = self.fc2(x)
        return F.log_softmax(x, dim=-1)


def get_data_loaders(train_batch_size, val_batch_size):
    data_transform = Compose([ToTensor(), Normalize((0.1307,), (0.3081,))])

    train_loader = DataLoader(MNIST(download=True, root=".", transform=data_transform, train=True),
                              batch_size=train_batch_size, shuffle=True)

    val_loader = DataLoader(MNIST(download=False, root=".", transform=data_transform, train=False),
                            batch_size=val_batch_size, shuffle=False)
    return train_loader, val_loader


def run(train_batch_size, val_batch_size, epochs, lr, momentum, log_interval):
    train_loader, val_loader = get_data_loaders(train_batch_size, val_batch_size)
    model = Net()
    device = 'cpu'

    if torch.cuda.is_available():
        device = 'cuda'

    optimizer = SGD(model.parameters(), lr=lr, momentum=momentum)
    trainer = create_supervised_trainer(model, optimizer, F.nll_loss, device=device)
    evaluator = create_supervised_evaluator(model,
                                            metrics={'accuracy': CategoricalAccuracy(),
                                                     'nll': Loss(F.nll_loss)},
                                            device=device)
    
    pbar = ProgressBar(engine=trainer, loader=train_loader, mode='iteration', log_interval=log_interval)

    @trainer.on(Events.ITERATION_COMPLETED)
    def log_training_loss(engine):
        iter = (engine.state.iteration - 1) % len(train_loader) + 1
#         if iter % log_interval == 0:
#             print("Epoch[{}] Iteration[{}/{}] Loss: {:.2f}"
#                   "".format(engine.state.epoch, iter, len(train_loader), engine.state.output))

    @trainer.on(Events.EPOCH_COMPLETED)
    def log_training_results(engine):
        evaluator.run(train_loader)
        metrics = evaluator.state.metrics
        avg_accuracy = metrics['accuracy']
        avg_nll = metrics['nll']
#         print("Training Results - Epoch: {}  Avg accuracy: {:.2f} Avg loss: {:.2f}"
#               .format(engine.state.epoch, avg_accuracy, avg_nll))

    @trainer.on(Events.EPOCH_COMPLETED)
    def log_validation_results(engine):
        evaluator.run(val_loader)
        metrics = evaluator.state.metrics
        avg_accuracy = metrics['accuracy']
        avg_nll = metrics['nll']
#         print("Validation Results - Epoch: {}  Avg accuracy: {:.2f} Avg loss: {:.2f}"
#               .format(engine.state.epoch, avg_accuracy, avg_nll))

    trainer.run(train_loader, max_epochs=epochs)
```
and output:
```
[0/938]   0%|           [00:00<?]

Iteration 10
Iteration 20
Iteration 30

[0/938]   0%|           [00:00<?]

Iteration 40
Iteration 50
Iteration 60

[0/938]   0%|           [00:00<?]

Iteration 70
Iteration 80
Iteration 90

[0/938]   0%|           [00:00<?]

Iteration 100
Iteration 110
Iteration 120

[0/938]   0%|           [00:01<?]

Iteration 130
Iteration 140
Iteration 150

[0/938]   0%|           [00:01<?]

Iteration 160
Iteration 170
Iteration 180

[0/938]   0%|           [00:01<?]

Iteration 190
Iteration 200
Iteration 210

[0/938]   0%|           [00:01<?]

Iteration 220
Iteration 230
Iteration 240

[0/938]   0%|           [00:01<?]

Iteration 250
Iteration 260
Iteration 270

[0/938]   0%|           [00:02<?]

Iteration 280
Iteration 290
Iteration 300
```

I would like to see in this case a plain progress barm just showing batch iterations without `tqdm.write`. What do you think ?

Reply:
you are still not adding the pbar to the trainer

add this line:
`trainer.add_handler(Events.ITERATION_COMPLETED, pbar)`

> I would like to see in this case a plain progress barm just showing batch iterations without tqdm.write. What do you think?

I could check if `log_interval` is `None`, and if so I won't output the `tqdm.write()` message. Would this do?

Reply:
> but for both modes, the triggered event should still be ITERATION_COMPLETED

If it is necessary part, you can add it directly in the `__init__` ?

Reply:
yes, I can, but is that good practice?